canonicalwebteam/discourse_docs/app.py

```diff
@@ -0,0 +1,35 @@
+from canonicalwebteam.discourse_docs.blueprint import build_blueprint
+from canonicalwebteam.discourse_docs.models import DiscourseApi
+
+
+class DiscourseExtension(object):
+    def __init__(
+        self,
+        app=None,
+        url_prefix="/docs",
+        discourse_url=None,
```

Comment:
Ah, so this is actually the entrypoint interface of the application. So my comment about injecting `DiscourseApi` directly should be applied here:

``` python3
    def __init__(
        self,
        app=None,
        model=None,
        ...
    )
```

Reply:
As discussed, we'll leave this as it is for now.

Reply:
So while I was trying to actually add this module to docs.snapcraft.io, I realised there is a good reason for inverting the interface here. In docs.snapcraft.io, for the time being, we will want to maintain their existing error handlers - e.g. [the 404 handler](https://github.com/canonical-websites/docs.snapcraft.io/blob/master/webapp/app.py#L66):

``` python3
@app.errorhandler(404)
def page_not_found(e):
    try:
        frontpage, nav_html = discourse.parse_frontpage()
    except NavigationParseError as nav_error:
        nav_html = f"<p>{str(nav_error)}</p>"

return flask.render_template("404.html", nav_html=nav_html), 404
```

Now I don't think this logic belongs in this module, but at the same time, for now, it's legitimate for the 404 page to want to get that context. So it's probably nicer to do:

``` python3
# app.py

discourse_api = DiscourseAPI("https://forum.snapcraft.io/", 3781, 15)
discourse.init_app(app, discourse_api, "/")

# ...

@app.errorhandler(404)
def page_not_found(e):
    try:
        frontpage, nav_html = discourse_api.parse_frontpage()
    except NavigationParseError as nav_error:
        nav_html = f"<p>{str(nav_error)}</p>"

    return flask.render_template("404.html", nav_html=nav_html), 404
```

I think this illustrates that the clarity shown by this 2-line interface (rather than 1-line) is valuable enough to suggest it should be the default structure.

Reply:
ok  it makes sense.